internal/build/imgsrc/ensure_builder.go

```diff
@@ -0,0 +1,325 @@
+package imgsrc
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/samber/lo"
+	"github.com/superfly/fly-go"
+	"github.com/superfly/fly-go/flaps"
+	"github.com/superfly/flyctl/internal/flapsutil"
+	"github.com/superfly/flyctl/internal/flyutil"
+	"github.com/superfly/flyctl/internal/haikunator"
+	"github.com/superfly/flyctl/internal/tracing"
+)
+
+func EnsureBuilder(ctx context.Context, org *fly.Organization, region string) (*fly.Machine, *fly.App, error) {
+	ctx, span := tracing.GetTracer().Start(ctx, "ensure_builder")
+	defer span.End()
+
+	builderApp := org.RemoteBuilderApp
+	if builderApp != nil {
+		flaps, err := flapsutil.NewClientWithOptions(ctx, flaps.NewClientOpts{
+			AppName: builderApp.Name,
+			// TOOD(billy) make a utility function for App -> AppCompact
+			AppCompact: &fly.AppCompact{
+				ID:       builderApp.ID,
+				Name:     builderApp.Name,
+				Status:   builderApp.Status,
+				Deployed: builderApp.Deployed,
+				Hostname: builderApp.Hostname,
+				AppURL:   builderApp.AppURL,
+				Organization: &fly.OrganizationBasic{
+					ID:       builderApp.Organization.ID,
+					Name:     builderApp.Organization.Name,
+					Slug:     builderApp.Organization.Slug,
+					RawSlug:  builderApp.Organization.RawSlug,
+					PaidPlan: builderApp.Organization.PaidPlan,
+				},
+				PlatformVersion: builderApp.PlatformVersion,
+				PostgresAppRole: builderApp.PostgresAppRole,
+			},
+			OrgSlug: builderApp.Organization.Slug,
+		})
+		if err != nil {
+			tracing.RecordError(span, err, "error creating flaps client")
+			return nil, nil, err
+		}
+		ctx = flapsutil.NewContextWithClient(ctx, flaps)
+	}
+
+	builderMachine, err := validateBuilder(ctx, builderApp)
+	if err == nil {
+		span.AddEvent("builder app already exists and is valid")
+		return builderMachine, builderApp, nil
+	}
+
+	var validateBuilderErr ValidateBuilderError
+	if !errors.As(err, &validateBuilderErr) {
+		return nil, nil, err
+	}
+
+	if validateBuilderErr != NoBuilderApp {
```

Comment:
I believe this block won't occur because `NoBuilderApp` is an instance of `ValidateBuilderError` which is checked with the `errors.As()` above, correct?

Reply:
i don't believe so. errors.As just checks if it's a ValidateBuilderError or not, and we only return before this check if it isn't a validation error.

https://go.dev/play/p/I5_iTNzBwiQ